src/compiler/tsbuildPublic.ts

```diff
@@ -1346,94 +1464,23 @@ namespace ts {
-        }
-
         // Container if no files are specified in the project
         if (!project.fileNames.length && !canJsonReportNoInputFiles(project.raw)) {
             return {
                 type: UpToDateStatusType.ContainerOnly
             };
         }
 
-        // Collect the expected outputs of this project
-        const outputs = getAllProjectOutputs(project, !host.useCaseSensitiveFileNames());
-
-        // Now see if all outputs are newer than the newest input
-        let oldestOutputFileName = "(none)";
-        let oldestOutputFileTime = maximumDate;
-        let newestOutputFileName = "(none)";
-        let newestOutputFileTime = minimumDate;
-        let missingOutputFileName: string | undefined;
-        let newestDeclarationFileContentChangedTime = minimumDate;
-        let isOutOfDateWithInputs = false;
-        if (!force) {
-            for (const output of outputs) {
-                // Output is missing; can stop checking
-                // Don't immediately return because we can still be upstream-blocked, which is a higher-priority status
-                if (!host.fileExists(output)) {
-                    missingOutputFileName = output;
-                    break;
-                }
-
-                const outputTime = getModifiedTime(host, output);
-                if (outputTime < oldestOutputFileTime) {
-                    oldestOutputFileTime = outputTime;
-                    oldestOutputFileName = output;
-                }
-
-                // If an output is older than the newest input, we can stop checking
-                // Don't immediately return because we can still be upstream-blocked, which is a higher-priority status
-                if (outputTime < newestInputFileTime) {
-                    isOutOfDateWithInputs = true;
-                    break;
-                }
-
-                if (outputTime > newestOutputFileTime) {
-                    newestOutputFileTime = outputTime;
-                    newestOutputFileName = output;
-                }
-
-                // Keep track of when the most recent time a .d.ts file was changed.
-                // In addition to file timestamps, we also keep track of when a .d.ts file
-                // had its file touched but not had its contents changed - this allows us
-                // to skip a downstream typecheck
-                if (isDeclarationFileName(output)) {
-                    const outputModifiedTime = getModifiedTime(host, output);
-                    newestDeclarationFileContentChangedTime = newer(newestDeclarationFileContentChangedTime, outputModifiedTime);
-                }
-            }
-        }
-
-        let pseudoUpToDate = false;
-        let usesPrepend = false;
-        let upstreamChangedProject: string | undefined;
+        // Fast check to see if reference projects are buildable
+        let referenceStatuses;
+        const force = !!state.options.force;
         if (project.projectReferences) {

```

Comment:
Moved up project references status check to see if we can build as the first check.. This ensures that we are not doing unnecessary input output etc checks in that case and can directly check buildInfo to do those checks and return.

Basically, anything that marks project as non buildable, is done first and then input output time checks. (except missing input time check as we build project anyways in that case)